sip/sip-002-smart-contract-language.md

```diff
@@ -0,0 +1,493 @@
+   transaction arguments, however, on initial release, we believe
+   dynamic invocation to be too dangerous to support.
+2. Called smart contracts _must_ exist at the time of creation.
+3. No cycles may exist in the call graph of a smart contract. This
+   prevents recursion (and re-entrancy bugs). Such structures can
+   be detected with static analysis of the call graph, and will be
+   rejected by the network.
+
+A key benefit of the static analyzability of this smart contracting
+language is that _all_ transactions that can possibly be called from
+the outermost transaction can be known _a priori_ so that a user can
+be warned about all side effects before signing a transaction.
+
+## Principals and Transaction Verification
+
+The language provides a primitive for checking whether or not the
+smart contract transaction was signed by a particular
+_principal_. Principals are a specific type in the smart contracting
+language which represent a signing entity (roughly equivalent to a
+Stacks address). The signature itself is not checked by the smart
+contract, but by the VM. A smart contract function can use a globally
+defined variable to obtain the current transaction's signer:
+
+```scheme
+tx-sender
+```
+
+Importantly, the `tx-sender` variable does not change during
+inter-contract calls. This means that if you call a transaction in a
+given smart contract, that transaction is able to make calls into
+other smart contracts on your behalf. This enables a wide variety of
+applications, but it comes with some dangers for users of smart
+contracts. However, as mentioned before, the static analysis
+guarantees of our smart contracting language allow clients to know a
+priori which transactions a given smart contract will ever call. Good
+clients should always warn users about any potential side effects
+of a given transaction.
+
+Note that most assets in the smart contracting language and blockchain
+will be "owned" by objects of the principal type, meaning that any
+object of the principal type may own an asset. For the case of
+public-key hash and multi-signature Stacks addresses, a given
+principal can operate on their assets by issuing a signed transaction
+on the blockchain. _Smart contracts_ may also be principals
+(reprepresented by the smart contract's identifier), however, there is
+no private key associated with the smart contract, and it cannot
+broadcast a signed transaction on the blockchain. In order to allow
+smart contracts to authorize transactions with the smart contract's
+own principal, smart contracts may use the special function:
+
+```scheme
+(contract-send-tx (...))
+```
+
+This function will execute the commands passed as an argument with the
+`tx-sender` set to the _contract's_ principal, rather than the current
+sender. It returns the return value of the provided commands. A smart
+contract may use the special variable `contract-principal` to refer to
+its own principal.
```

Comment:
Can you give an example of this?  I'm having a hard time appreciating the difference between this and `contract-call`.  Is the only difference that the "context" changes when the code runs? e.g. `tx-sender` (and possibly other values) change?

Reply:
Yep -- the main difference is the context change (`tx-sender`). Additionally, `contract-call` can only invoke a transaction function from the other smart contract. `contract-send-tx` (which is probably poorly named) takes arbitrary code as input, so it can do things like execute `stacks-transfer!` or call `contract-call` to call other transactions, etc.

Reply:
Gotcha.  Calling this kind of inter-contract interaction a "transaction" was really tripping me up.  Do you mean to say that there exists a notion of a first-class transaction between smart contracts?  Or are you really just trying to find a way to define a way to run a closure under the contract's principal instead of the principle who sent the client-submitted transaction?  If the latter, then can we call this something like `as-contract` instead?

Reply:
Either way, we might want to add some helper methods to help the code distinguish between a `principal` that is a user and a `principal` that is a contract.  For example, I might want it to be the case that only users can claim tokens this way.

Reply:
Yeah -- I like `as-contract`

Reply:
I updated it to use `as-contract` and added that example.

Reply:
Forgot to ask earlier.  What's `deftx` and why/how is it different from `defun`?

Reply:
`deftx` defines a transaction -- that's a function that can be called by issuing a transaction on the blockchain, or by another smart contract.

`define` defines a function, which is basically just a syntactic nicety -- the only legal functions defined this way are those able to inlined.

Reply:
Would also add that function defined with `define` are private -- adding that to the SIP, because I don't think that's clear in it.

Reply:
Gotcha, thanks for clarifying!